### DISCUSSION TITLE
How to pass params when using BigQueryInsertJobOperator

### DISCUSSION THREAD

**Discussion by 2stash** [apache/airflow#16531] [Q&A] [answered]
Hi! I posted this on stack overflow, but thought I might get more responses here.

I am trying to use BigQueryInsertJobOperator for a query of Google BigQuery. I am trying to pass variable names into a sql file, but I am getting an error. The error is slightly different based on what I type in the sql file, but it doesn't recognize the variable.

I believe Airflow is sending me the error and not google. Per google dogs params should be marked with an @, but airflow says unexpected symbol.

*** Dag and sql file have been simplified for brevity
My Dag

`TABLE_1 = "my_table"

with DAG(
    dag_id,
    schedule_interval='@daily',
    start_date= days_ago(1),
    template_searchpath = '/mypath/sql/',
    catchup = False,
   
    
) as dag:

    query_one = BigQueryInsertJobOperator(
        task_id='query_test',
        configuration={
            "query":{
                "query": "{% include 'big_insert_job_test.sql' %}",
                "defaultDataset":{
                    "datasetId":"my_dataset_id",
                    "projectId":"my_project_id"
                },
    
                "queryParameters": [
                    {
                    "name": "TABLE_1",
                    "parameterType": {
                        "type": "STRING",
                    },
                    "parameterValue": {"value":TABLE_1}
                    }
                ]   
    
                            
            }
        }`

SQL File

`SELECT * FROM @TABLE_1`

According to the google docs I can pass a variable as a "@" or a "?". I have tried sending the variables many different ways:

`SELECT * FROM {{params.TABLE_1}}. // This is how it works with BigQueryExecuteQueryOperator, 
but that is deprecated so trying to use the BigQueryInsertJobOperator

SELECT * FROM `TABLE_1` // just an example, 
I have tried every combo of "", '', {}, etc`

Stackoverflow Link
https://stackoverflow.com/questions/68027380/how-do-you-pass-variables-with-bigqueryinsertjoboperator-in-airflow

**Comment by 2stash** [comment, marked as answer]
I figured it out. I don't see this discussed in any of the docs, but there is an example, that I looked at many times, before I figured out the example actually gave me an answer.

Short answer is, use user_defined_macros={"DATASET": DATASET_NAME, "TABLE": TABLE_1}. This is set in the DAG config area, and not down in the operators.

I posted an answer on stackoverflow as well.

https://stackoverflow.com/questions/68027380/how-do-you-pass-variables-with-bigqueryinsertjoboperator-in-airflow/68039842#68039842

**Comment by cwdjankoski** [comment]
@2stash Thank you for taking the time to come back and post the correct approach 🙇. After 2 days of fighting with the `queryParameters` from inside the `configuration` arg - mentioned in the official documentation (!) - i was able to make it work thanks to your comments.  

What would be good approach ? 
 - adjust the documenation to mention this approach ?
 - or maybe make a more explicit example in the official docs on how to use the `queryParameters`  

What do you think ? I think this is a common use-case and can save potentially many people a lot of troubles and time

**Reply by potiuk** [reply]
PRs are always very welcome. Bith docs and examples are cool. go for both.

**Reply by gurpreetjhu** [reply]
@cwdjankoski David--not sure you remember this on top of your head, but I am struggling as I have 1 SQL which is being called multiple times , just one parameter is getting changed. Some how post following the Stackoverflow my logic still not working ..airflow say varibale undefined.

Can you please point some document..

**Comment by messam88** [comment]
Sorry for posting this response a bit late but i was tried it now 
Following the same approach i used it with BigQueryInsertJobOperator 
```
"query": f"{{% include {file_name} %}}",
"queryParameters": [
                    {
                        "name": "custom_start_date",
                        "parameterType": {"type": "STRING"},
                        "parameterValue": {"value": "{{ macros.ds_add(ds, -10) }}"}
                    },
                    {
                        "name": "custom_end_date",
                        "parameterType": {"type": "STRING"},
                        "parameterValue": {"value": "{{ ds }}"}
                    },
                ],
```
But to access the parameter inside the sql file you have to use it like that 
`@custom_start_date AND @custom_end_date`
Hope this might be helpful to anyone
